### DISCUSSION TITLE
Create an offer -> Invalid order signature

### DISCUSSION THREAD

**Discussion by ServantMaximus** [ProjectOpenSea/seaport#643] [Q&A] [answered]
Hi,

I am trying to use the https://api.opensea.io/v2/orders/ethereum/seaport/offers endpoint to create an offer.
Docs: https://docs.opensea.io/v2.0/reference/create-an-offer

3 parameters need to be fulfilled:
- Parameters: I create json input (=message) that is correct. The value of the msg variable in the code below is the input for this field.
- Signature: I sign my json input (=message) using the code below and fill in the signature in this field
- API Key ✅

However I get the error: 
"errors": [
    "['Invalid order signature']"
  ]

**I am a bit clueless on why my signature is invalid?**

My Python code to generate the signature:

```
from web3.auto import w3
from eth_account.messages import encode_defunct
from hexbytes import HexBytes
import requests
import json

msg = '''{"offerer":"0x709e101d81A76B38800F5fE4E390CC32b7c854a7","offer":[{"itemType":1,"token":"0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2","identifierOrCriteria":"0","startAmount":"10000000000000000","endAmount":"10000000000000000"}],"consideration":[{"itemType":2,"token":"0xF87E31492Faf9A91B02Ee0dEAAd50d51d56D5d4d","identifierOrCriteria":"36410213260540415590581082995199198625690","startAmount":"1","endAmount":"1","recipient":"0x709e101d81A76B38800F5fE4E390CC32b7c854a7"},{"itemType":1,"token":"0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2","identifierOrCriteria":"0","startAmount":"250000000000000","endAmount":"250000000000000","recipient":"0x9A6ebE7E2a7722F8200d0ffB63a1F6406A0d7dce"}],"totalOriginalConsiderationItems":2,"startTime":"1661679182","endTime":"1661765694","orderType":2,"zone":"0x004C00500000aD104D7DBd00e3ae0A5C00560C00","zoneHash":"0x0000000000000000000000000000000000000000000000000000000000000000","salt":5578881512597576,"conduitKey":"0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000","counter":0}'''

private_key = "0xprivate key"
message = encode_defunct(text=msg)
signed_message =  w3.eth.account.sign_message(message, private_key=private_key)
signature = HexBytes.hex(signed_message.signature)

print(signature)

```

**Comment by ServantMaximus** [comment]
Additionally: 

I signed the message below using etherscan:
`{"offerer":"0x709e101d81A76B38800F5fE4E390CC32b7c854a7",
    "offer":
        [{"itemType":1,"token":"0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2","identifierOrCriteria":"0","startAmount":"10000000000000000","endAmount":"10000000000000000"}],
    "consideration":[
        {"itemType":2,"token":"0xF87E31492Faf9A91B02Ee0dEAAd50d51d56D5d4d","identifierOrCriteria":"22458636216781938588582724090496701955961","startAmount":"1","endAmount":"1","recipient":"0x709e101d81A76B38800F5fE4E390CC32b7c854a7"},
        {"itemType":1,"token":"0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2","identifierOrCriteria":"0","startAmount":"250000000000000","endAmount":"250000000000000","recipient":"0x9A6ebE7E2a7722F8200d0ffB63a1F6406A0d7dce"}],
    "totalOriginalConsiderationItems":2,
    "startTime":"1661967100",
    "endTime":"1661969099",
    "orderType":2,
    "zone":"0x004C00500000aD104D7DBd00e3ae0A5C00560C00",
    "zoneHash":"0x0000000000000000000000000000000000000000000000000000000000000000",
    "salt":5578881512597576,
    "conduitKey":"0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000",
    "counter":0}`

I generate the signature on https://etherscan.io/verifiedSignatures#.
![image](https://user-images.githubusercontent.com/110773480/187046677-35863017-0582-4a1d-86ff-877bcbfee85e.png)

I than used the generated signature to test using the browser, still not working:
The generated signature by etherscan corresponds with the signature that my Python code generates, so I assume this should be a correct signature...
![image](https://user-images.githubusercontent.com/110773480/187046752-ccdf76be-78cf-4339-ae59-ce0528e4622c.png)

This generated the following Python code:
`import requests

url = "https://api.opensea.io/v2/orders/ethereum/seaport/offers"

payload = {
    "parameters": {
        "offerer": "0x709e101d81A76B38800F5fE4E390CC32b7c854a7",
        "offer": [
            {
                "itemType": 1,
                "token": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
                "identifierOrCriteria": "0",
                "startAmount": "10000000000000000",
                "endAmount": "10000000000000000"
            }
        ],
        "consideration": [
            {
                "itemType": 2,
                "token": "0xF87E31492Faf9A91B02Ee0dEAAd50d51d56D5d4d",
                "identifierOrCriteria": "22458636216781938588582724090496701955961",
                "startAmount": "1",
                "endAmount": "1",
                "recipient": "0x709e101d81A76B38800F5fE4E390CC32b7c854a7"
            },
            {
                "itemType": 1,
                "token": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
                "identifierOrCriteria": "0",
                "startAmount": "250000000000000",
                "endAmount": "250000000000000",
                "recipient": "0x9A6ebE7E2a7722F8200d0ffB63a1F6406A0d7dce"
            }
        ],
        "totalOriginalConsiderationItems": 2,
        "startTime": "1661967100",
        "endTime": "1661969099",
        "orderType": 2,
        "zone": "0x004C00500000aD104D7DBd00e3ae0A5C00560C00",
        "zoneHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
        "salt": 5578881512597576,
        "conduitKey": "0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000",
        "counter": 0
    },
    "signature": "0x663ed43b9ee8d02bed381d4381ebc03bd6040f347b531dc0d12fb5d95f5e57132d7676a0207a45ad5c89bed2b3216848119873c4e576e17d173233b371c4b64b1b"
}
headers = {
    "Accept": "application/json",
    "X-API-KEY": "My Opensea API KEY",
    "Content-Type": "application/json"
}

response = requests.post(url, json=payload, headers=headers)

print(response.text)`

**Comment by malihdabboussi** [comment]
i am getting the same error message when doing listing

**Reply by onessss** [reply]
have you resolved it？

**Comment by ryanio** [comment]
You have to sign the order using EIP-712

**Reply by AliMurtazaJaffery** [reply]
Hi! can you share the code in Python for the signature part?
\

**Reply by AliMurtazaJaffery** [reply]
@ryanio

**Comment by granolaboy** [comment, marked as answer]
Hi everyone!

Was having the same problem trying to produce a valid EIP-712 signature to create a new listing in python.

My issue was that I was using old eth libs. `encode_structured_data` is **bugged in older versions** (in particular with arrays). 

do this:

`pip list --outdated`

and if you see anything with "web3" or "eth" in its name - you need to update your libs to the latest versions. I had to remove brownie and eip712 due to version conflicts.

This is what mine looks like in the end (with the dict pretty-printed and pasted in. You can figure out how to make that dict yourself.

I'm listing a Mori Mori (contract `0x85203d271c09ea3d76688753B6f616aEF1e09B12` token id `248` erc-721) for 0.1 ETH

```
    message = {'domain': {'chainId': 1,
            'name': 'Seaport',
            'verifyingContract': '0x00000000006c3852cbEf3e08E8dF289169EdE581',
            'version': '1.1'},
 'message': {'conduitKey': b'\x00\x00\x00{\x02#\x00\x91\xa7\xed\x01#'
                           b'\x00r\xf7\x00j\x00M`\xa8\xd4\xe7\x1dY\x9b\x81\x04'
                           b'%\x0f\x00\x00',
             'consideration': [{'endAmount': 92000000000000000,
                                'identifierOrCriteria': 0,
                                'itemType': 0,
                                'recipient': '0x0063E887827B5c5f531d0C9559478842c57Ef7Ab',
                                'startAmount': 92000000000000000,
                                'token': '0x0000000000000000000000000000000000000000'},
                               {'endAmount': 2500000000000000,
                                'identifierOrCriteria': 0,
                                'itemType': 0,
                                'recipient': '0x0000a26b00c1F0DF003000390027140000fAa719',
                                'startAmount': 2500000000000000,
                                'token': '0x0000000000000000000000000000000000000000'},
                               {'endAmount': 5500000000000000,
                                'identifierOrCriteria': 0,
                                'itemType': 0,
                                'recipient': '0xA81cA450D7ea11efEe17d6b4A8f743B3b6AE46e0',
                                'startAmount': 5500000000000000,
                                'token': '0x0000000000000000000000000000000000000000'}],
             'counter': 1,
             'endTime': 1674505418,
             'offer': [{'endAmount': 1,
                        'identifierOrCriteria': 248,
                        'itemType': 2,
                        'startAmount': 1,
                        'token': '0x85203d271c09ea3d76688753B6f616aEF1e09B12'}],
             'offerer': '0x0063e887827b5c5f531d0c9559478842c57ef7ab',
             'orderType': 2,
             'salt': 94632875894598549038100692318865073957,
             'startTime': 1674504218,
             'zone': '0x004c00500000ad104d7dbd00e3ae0a5c00560c00',
             'zoneHash': b'1\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                         b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
                         b'\x00\x00\x00\x00\x00\x00\x00\x00'},
 'primaryType': 'OrderComponents',
 'types': {'ConsiderationItem': [{'name': 'itemType', 'type': 'uint8'},
                                 {'name': 'token', 'type': 'address'},
                                 {'name': 'identifierOrCriteria',
                                  'type': 'uint256'},
                                 {'name': 'startAmount', 'type': 'uint256'},
                                 {'name': 'endAmount', 'type': 'uint256'},
                                 {'name': 'recipient', 'type': 'address'}],
           'EIP712Domain': [{'name': 'name', 'type': 'string'},
                            {'name': 'version', 'type': 'string'},
                            {'name': 'chainId', 'type': 'uint256'},
                            {'name': 'verifyingContract', 'type': 'address'}],
           'OfferItem': [{'name': 'itemType', 'type': 'uint8'},
                         {'name': 'token', 'type': 'address'},
                         {'name': 'identifierOrCriteria', 'type': 'uint256'},
                         {'name': 'startAmount', 'type': 'uint256'},
                         {'name': 'endAmount', 'type': 'uint256'}],
           'OrderComponents': [{'name': 'offerer', 'type': 'address'},
                               {'name': 'zone', 'type': 'address'},
                               {'name': 'offer', 'type': 'OfferItem[]'},
                               {'name': 'consideration',
                                'type': 'ConsiderationItem[]'},
                               {'name': 'orderType', 'type': 'uint8'},
                               {'name': 'startTime', 'type': 'uint256'},
                               {'name': 'endTime', 'type': 'uint256'},
                               {'name': 'zoneHash', 'type': 'bytes32'},
                               {'name': 'salt', 'type': 'uint256'},
                               {'name': 'conduitKey', 'type': 'bytes32'},
                               {'name': 'counter', 'type': 'uint256'}]}}

    signable_message = encode_structured_data(message)
    signed_message = web3.eth.account.sign_message(signable_message, private_key)
    signature = signed_message['signature'].hex()
```

**Reply by granolaboy** [reply]
https://github.com/ConsenSysMesh/py-eip712-structs may be of help to you. You are creating an EIP-712 signature.

You have to get the consideration items from the contracts endpoint.

**Reply by hao0707** [reply]
> https://github.com/ConsenSysMesh/py-eip712-structs may be of help to you. You are creating an EIP-712 signature.
> 
> You have to get the consideration items from the contracts endpoint.
Hi,have you successfully placed an order through the opensea api using this method?

**Reply by onessss** [reply]
> > https://github.com/ConsenSysMesh/py-eip712-structs may be of help to you. You are creating an EIP-712 signature.
> > You have to get the consideration items from the contracts endpoint.
> > Hi,have you successfully placed an order through the opensea api using this method?

i have resolve this issues,thank ur reply

**Reply by hao0707** [reply]
> > > https://github.com/ConsenSysMesh/py-eip712-structs may be of help to you. You are creating an EIP-712 signature.
> > > You have to get the consideration items from the contracts endpoint.
> > > Hi,have you successfully placed an order through the opensea api using this method?
> 
> i have resolve this issues,thank ur reply

Hi,can you show your code？I have been troubled by this for a long time,thank you very much!
